jcl/src/java.base/share/classes/java/lang/StackTraceElement.java

```diff
@@ -243,7 +307,18 @@ public boolean isNativeMethod() {
 @Override
 public String toString() {
 	StringBuilder buf = new StringBuilder(80);
-	Util.printStackTraceElement(this, source, buf, false);
+	/**
+	 * includeExtendedInfo is essentially a check to see if this StackTraceElement
+	 * was created using a public constructor. Both includeClassLoaderName and
+	 * includeModuleVersion are initialized to true in the public constructors,
+	 * since these pieces of information are to be included in the stack trace.
+	 *
+	 * It's also possible that both of these flags are set to true natively,
+	 * in which case we still want to include extended info when printing the
+	 * stack trace.
+	 */
+	boolean includeExtendedInfo = includeClassLoaderName && includeModuleVersion;
```

Comment:
This is an optimization? I don't see that changing this makes any difference to the behavior.

Reply:
This is not an optimization. It is needed because both public constructor and non-public constructor `StackTraceElement` objects may call `StackTraceElement.toString()`, and `includeExtendedInfo` allows us to indicate to `printStackTraceElement()` whether extended info (i.e. `classLoaderName` and `moduleVersion`) should be included or not.

Calls to `printStackTraceElement()` need to indicate if the extended info should be included or not.

Reply:
Looking at it again, I think it does do some extra work, since `includeClassLoaderName()`/`includeModuleVersion()` will cover the check. So this can be removed.

Reply:
However, including `includeExtendedInfo` will make the call to `printStackTraceElement()` make more sense, since it will set `includeExtendedInfo` to the right value. Passing `false` may be confusing since the extended info gets included after all, if includeClassLoaderName()`/`includeModuleVersion()` end up being true.

Setting it does allow us to skip calling `includeClassLoaderName()`/`includeModuleVersion()` in `Util.printStackTraceElement()` if it is true.